src/main/scala/mesosphere/mesos/TaskBuilder.scala

```diff
@@ -342,33 +346,83 @@ object TaskBuilder {
     builder.build()
   }
 
-  def portsEnv(definedPorts: Seq[Int], assignedPorts: Seq[Int], portNames: Seq[Option[String]]): Map[String, String] = {
+  // portsEnv generates $PORT{x} and $PORT_{y} environment variables, wherein `x` is an index into
+  // the portDefinitions or portMappings array and `y` is a non-zero port specifically requested by
+  // the application specification.
+  //
+  // @param definedPorts are either declared container ports (if port mappings are specified) or host ports; may be 0's
+  // @param assignedPorts are "real" host ports allocated from Mesos resource offers
+  // @return a dictionary of variables that should be added to a tasks environment
+  def portsEnv(
+    definedPorts: Seq[Int],
+    assignedPorts: Seq[Option[Int]],
+    portNames: Seq[Option[String]]): Map[String, String] = {
     if (assignedPorts.isEmpty) {
       Map.empty
     }
     else {
       val env = Map.newBuilder[String, String]
+      val generatedPorts = Map.newBuilder[Int, Int] // index -> container port
+
+      object ContainerPortGenerator {
+        // track which port numbers are already referenced by PORT_xxx envvars
+        lazy val consumedPorts = scala.collection.mutable.Set(definedPorts: _*) ++= assignedPorts.flatten
+        lazy val rand = new scala.util.Random(System.currentTimeMillis)
```

Comment:
Probably want to just use "Random" instead of seeding it (it should be seeded automatically)